doc: Adds doc for implementation of csi-addons (#9624)
rook/rook: Documentation/ceph-csi-drivers.md

The Volume Replication Operator is a kubernetes operator that provides common and reusable APIs for storage disaster recovery. It is based on [csi-addons/spec](https://github.com/csi-addons/spec) specification and can be used by any storage provider.
It follows the controller pattern and provides extended APIs for storage disaster recovery. The extended APIs are provided via Custom Resource Definitions (CRDs).

> ### Prerequisites

Reviewer: In this position, it appears K8s 1.21 is only a prereq for RBD mirroring. Is that the intention? Or is there any K8s requirement for the CSI addons?

> ### Prerequisites

Author: Yes, it is not related to csi-addons and hence, moved it up. Earlier it was part of RBD mirroring so place it there.

## CSI-Addons Controller

The CSI-Addons Controller handles the requests from users to initiate an operation. Users create a CR that the controller inspects, and forwards a request to one or more CSI-Addons side-cars for execution.

### Enable the CSI-Addons Sidecar

To use the features provided by the CSI-Addons, the `csi-addons`
containers need to be deployed in the RBD provisioner and nodeplugin pods,
which are not enabled by default.

Execute the following command in the cluster to enable the CSI-Addons
sidecar:

* Edit the `rook-ceph-operator-config` configmap and add the
 following configurations

```bash
kubectl edit cm rook-ceph-operator-config -n rook-ceph
```

* Add the following properties if not present:

```yaml
data:
  CSI_ENABLE_CSIADDONS: "true"
```

* After updating the configmap with those settings, a new sidecar
 should now start automatically in the CSI provisioner and nodeplugin pod.

> NOTE: Make sure the version of ceph-csi used is v3.5.0+

> ### Deploying the controller

Reviewer: The controller should be deployed before the sidecar, right? Otherwise the sidecar at least fails if the CRDs aren't there. So this section should be first?

Author: yes. Ack!

Users can deploy the controller by running the following command:

```bash

> kubectl create -f https://github.com/csi-addons/kubernetes-csi-addons/deploy/controller/setup-controller.yaml

Reviewer: That URL doesn't work yet, it will be something like this after it merges:
```suggestion
kubectl create -f https://github.com/csi-addons/kubernetes-csi-addons/blob/main/deploy/controller/setup-controller.yaml
```
We should also link a versioned url instead of just main, otherwise it's difficult to know what is installed.

Author: yes, for the url to work the PR mentioned earlier needs to be merged first.

Reviewer 2: There is a release from the kubernetes-csi-addons project now that provides generated yaml files as assets:
- https://github.com/csi-addons/kubernetes-csi-addons/releases/tag/v0.3.0

That means documentation on installing the csi-addons-controller can refer to `https://github.com/csi-addons/kubernetes-csi-addons/releases/download/v0.3.0/setup-controller.yaml` and other yaml files that can be applied with `kubectl`.

```

This creates the required crds and configure permissions.

> ```bash

Reviewer: No need for this output in the doc, how about removing it?

Author: okay

clusterrolebinding.rbac.authorization.k8s.io/csi-addons-proxy-rolebinding created
configmap/csi-addons-manager-config created
service/csi-addons-controller-manager-metrics-service created
deployment.apps/csi-addons-controller-manager created
```

### Reclaim Space Operation

If a user wants to initiate a Reclaim Space operation, he will create a ReclaimSpaceJob
CR that includes the reclaim space as an operation.

```yaml
apiVersion: csiaddons.openshift.io/v1alpha1
kind: ReclaimSpaceJob
metadata:
  name: sample-1
spec:
  target:
    persistentVolumeClaim: pvc-1
  backOffLimit: 10
  retryDeadlineSeconds: 900
```

A user that created the ReclaimSpaceJob can follow the progress, and
check the result of the operation:

```bash
kubectl apply -f reclaimspace.yaml
job.csiaddons.openshift.io/sample-1 created
kubectl get reclaimspacejobs
NAME                       COMPLETIONS   DURATION   AGE   SPACE RECLAIMED
sample-1               0/1           6s         6s    -
[admin@ocs]$ kubectl get reclaimspacejobs
NAME                       COMPLETIONS   DURATION   AGE   SPACE RECLAIMED
sample-1               1/1           10s        25s   153mb
```

> ### ReclaimSpaceCronJob

Reviewer: Is there documentation for the csi addons in that repo? If so, it would make more sense for just mentioning basic features of the csi addons here in the rook docs, and link to the csi addons repo for more details. This seems like a lot of detail for the rook docs.

Author: okay sure.

```yaml
apiVersion: csiaddons.openshift.io/v1alpha1
kind: ReclaimSpaceCronJob
metadata:
  name: reclaimspacecronjob-sample
spec:
  concurrencyPolicy: Forbid
  failedJobsHistoryLimit: 1
  jobTemplate:
    spec:
      backOffLimit: 6
      retryDeadlineSeconds: 600
      target:
        persistentVolumeClaim: data-pvc
  schedule: '@weekly'
  successfulJobsHistoryLimit: 3
```

> ### Annotating PerstentVolumeClaims

Reviewer: ```suggestion
### Annotating PersistentVolumeClaims
```

persistentvolumeclaim/data-pvc annotated

$ kubectl get reclaimspacecronjobs.csiaddons.openshift.io 
NAME                    SCHEDULE    SUSPEND   ACTIVE   LASTSCHEDULE   AGE
data-pvc-1642663516   @midnight                                     3s

$ kubectl annotate pvc data-pvc "reclaimspace.csiaddons.openshift.io/schedule=*/1 * * * *" --overwrite=true
persistentvolumeclaim/data-pvc annotated

$ kubectl get reclaimspacecronjobs.csiaddons.openshift.io
NAME                  SCHEDULE    SUSPEND   ACTIVE   LASTSCHEDULE   AGE
data-pvc-1642664617   */1 * * * *                                   3s
```

* Upon adding annotation as shown above, a `ReclaimSpaceCronJob` with name
  `"<pvc-name>-xxxxxxx"` is created (pvc name suffixed with current time hash when 
  the job was created). 
* `schedule` value is in the same [format as Kubernetes CronJobs][batch_cronjob]
  that sets the and/or interval of the recurring operation request.
* Default schedule value `"@weekly"` is used if `schedule` value is empty or in invalid format. 
* `ReclaimSpaceCronJob` is recreated when `schedule` is modified and deleted when
  the annotation is removed.

> ### Prerequisites

Reviewer: Are these the prereqs for the csi addons, or for the previous section? If each section has a different version requirement, how about we include at the top of each section instead of at the end?

Author: not for csi-addons, will update it.

Refer to [ephemeral-doc](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes) for more info.
Also, See the example manifests for an [RBD ephemeral volume](https://github.com/rook/rook/tree/{{ branchName }}/deploy/examples/csi/rbd/pod-ephemeral.yaml) and a [CephFS ephemeral volume](https://github.com/rook/rook/tree/{{ branchName }}/deploy/examples/csi/cephfs/pod-ephemeral.yaml).

## CSI-Addons Controller

The CSI-Addons Controller handles the requests from users to initiate an operation. Users create a CR that the controller inspects, and forwards a request to one or more CSI-Addons side-cars for execution.
By listing the CSIAddonsNode CRs, the CSI-Addons Controller knows how to connect to the side-cars. By checking the supported capabilities of the side-cars, it can decide where to execute operations that the user requested.

### Installation by operator

> A CSI-Addons bundle can be used to install the CSI-Addons Controller with the following steps:

Reviewer: Installing the bundle is a black box to the rook users. How about a link to the csiaddons that tells more details about what will be created in the namespace?

```bash
$ kubectl create namespace storage-csi-addons

> $ operator-sdk run bundle -n storage-csi-addons quay.io/csiaddons/k8s-bundle:latest

Reviewer: The `latest` tag may not always be stable. Is it based off main or the latest stable release?

> $ operator-sdk run bundle -n storage-csi-addons quay.io/csiaddons/k8s-bundle:latest

Reviewer 3: not everyone will have operator-sdk installed. we should point to the doc on installing with just kubectl also

```

### Enable the CSI-Addons Sidecar

To use the features provided by the CSI-Addons, the `csi-addons`
containers need to be deployed in the RBD provisioner and nodeplugin pods,
which are not enabled by default.

Execute the following command in the cluster to enable the CSI-Addons
sidecar:

* Edit the `rook-ceph-operator-config` configmap and add the
 following configurations

```bash

> kubectl edit cm rook-ceph-operator-config -n rook-ceph

Reviewer 4: ig You can directly use `Kubectl patch`  here for updating the spec 
Doc: https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/

```

* Add the following properties if not present:

```yaml
data:
  CSI_ENABLE_CSIADDONS: "true"
```

* After updating the configmap with those settings, a new sidecar
 should now start automatically in the CSI provisioner and nodeplugin pod.

> NOTE: Make sure the version of ceph-csi used is v3.5.0+

Author: done

### Reclaim Space Operation

If a user wants to initiate a Reclaim Space operation, he will create a ReclaimSpaceJob
CR that includes the reclaim space as an operation.

```yaml
apiVersion: csiaddons.openshift.io/v1alpha1

> kind: ReclaimSpaceJob

Reviewer 3: cronjob or annotation is most useful for users can be documented that example?

Author: done

metadata:
  name: sample-1
spec:
  target:
    persistentVolumeClaim: pvc-1
  backOffLimit: 10
  retryDeadlineSeconds: 900
```
A user that created the ReclaimSpaceJob can follow the progress, and
check the result of the operation:

```bash

> [admin@ocs]$ kubectl apply -f reclaimspace.yaml

Reviewer 4: plz dont mention `[admin@ocs]$`

Author: done

To use the features provided by the CSI-Addons, the `csi-addons`
containers need to be deployed in the RBD provisioner and nodeplugin pods,
which are not enabled by default.

Execute the following command in the cluster to enable the CSI-Addons
sidecar:

* Update the `rook-ceph-operator-config` configmap and patch the
 following configurations

```bash
kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_ENABLE_CSIADDONS": "true"'
```

* After updating the configmap with those settings, a new sidecar
 should now start automatically in the CSI provisioner and nodeplugin pod.

> NOTE: Make sure the version of ceph-csi used is v3.5.0+

### CSI-ADDONS Operation

CSI-Addons supports the following operations:

- ReclaimSpace
  - [Creating ReclaimSpaceJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#reclaimspacejob)
  - [Creating ReclaimSpaceCronJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#reclaimspacecronjob)
  - [Annotating PersistentVolumeClaims](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#annotating-perstentvolumeclaims)

> - [NetworkFence](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md)

Reviewer 5: ```suggestion
- NetworkFence
  - [Creating a NetworkFence](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md)
```

### Deploying the controller

Users can deploy the controller by running the following commands:

```bash
kubectl create -f https://raw.githubusercontent.com/csi-addons/kubernetes-csi-addons/v0.3.0/deploy/controller/crds.yaml
kubectl create -f https://raw.githubusercontent.com/csi-addons/kubernetes-csi-addons/v0.3.0/deploy/controller/rbac.yaml
kubectl create -f https://raw.githubusercontent.com/csi-addons/kubernetes-csi-addons/v0.3.0/deploy/controller/setup-controller.yaml
```

This creates the required crds and configure permissions.

### Enable the CSI-Addons Sidecar

To use the features provided by the CSI-Addons, the `csi-addons`
containers need to be deployed in the RBD provisioner and nodeplugin pods,
which are not enabled by default.

Execute the following command in the cluster to enable the CSI-Addons
sidecar:

* Update the `rook-ceph-operator-config` configmap and patch the
 following configurations

```bash
kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_ENABLE_CSIADDONS": "true"'
```

> * After updating the configmap with those settings, a new sidecar

Reviewer 3: ```suggestion
* After enabling `CSI_ENABLE_CSIADDONS` in the configmap, a new sidecar container with name `csi-addons`
```

>  should now start automatically in the CSI provisioner and nodeplugin pod.

Reviewer 3: ```suggestion
 should now start automatically in the RBD CSI provisioner and nodeplugin pods.
```

> NOTE: Make sure the version of ceph-csi used is v3.5.0+

### CSI-ADDONS Operation

CSI-Addons supports the following operations:

> - ReclaimSpace

Reviewer: This isn't referring to a specific type, right? How about a more readable bullet such as:
```suggestion
- Reclaim Space
```

Author: Everywhere else it is used as a single word and I would prefer it to be the same. Let me know if you still want me to change this. @travisn

  - [Creating ReclaimSpaceJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#reclaimspacejob)
  - [Creating ReclaimSpaceCronJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#reclaimspacecronjob)
  - [Annotating PersistentVolumeClaims](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#annotating-perstentvolumeclaims)

> - NetworkFence

Reviewer: ```suggestion
- Network Fencing
```

Author: NetworkFence is a kind so I guess it is fine this way.

>   - [Creating a ReclaimSpaceJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#reclaimspacejob)
>   - [Creating a ReclaimSpaceCronJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#reclaimspacecronjob)
>   - [Annotating PersistentVolumeClaims](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/reclaimspace.md#annotating-perstentvolumeclaims)
> - Network Fencing
>   - [Creating a NetworkFence](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md)

Reviewer 6: ```suggestion
  - [Creating a ReclaimSpaceJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/v0.3.0/docs/reclaimspace.md#reclaimspacejob)
  - [Creating a ReclaimSpaceCronJob](https://github.com/csi-addons/kubernetes-csi-addons/blob/v0.3.0/docs/reclaimspace.md#reclaimspacecronjob)
  - [Annotating PersistentVolumeClaims](https://github.com/csi-addons/kubernetes-csi-addons/blob/v0.3.0/docs/reclaimspace.md#annotating-perstentvolumeclaims)
- Network Fencing
  - [Creating a NetworkFence](https://github.com/csi-addons/kubernetes-csi-addons/blob/v0.3.0/docs/networkfence.md)
```
